consensus/istanbul/backend/handler.go

```diff
@@ -227,39 +227,111 @@ func (sb *Backend) SetP2PServer(p2pserver consensus.P2PServer) {
 
 // This function is called by miner/worker.go whenever it's mainLoop gets a newWork event.
 func (sb *Backend) NewWork() error {
+	sb.logger.Debug("Posting FinalCommittedEvent 1", "func", "NewWork")
+
 	sb.coreMu.RLock()
 	defer sb.coreMu.RUnlock()
 	if !sb.coreStarted {
 		return istanbul.ErrStoppedEngine
 	}
 
+	sb.logger.Debug("Posting FinalCommittedEvent 2", "func", "NewWork")
+
 	go sb.istanbulEventMux.Post(istanbul.FinalCommittedEvent{})
 	return nil
 }
 
-// This function is called by all nodes.
-// At the end of each epoch, this function will
-//    1)  Output if it is or isn't an elected validator if it has mining turned on.
-//    2)  Refresh the validator connections if it's a proxy or non proxied validator
+// Determine if this validator signed the parent of the supplied block:
+// First check the grandparent's validator set. If not elected, it didn't.
+// Then, check the parent seal on the supplied block.
+// We cannot determine any specific info from the validators in the seal of
+// the parent block, because different nodes circulate different versions.
+// It only becomes canonical when the child block is proposed.
+func (sb *Backend) ValidatorElectedAndSignedParentBlock(child *types.Block) (elected bool, inParentSeal bool, countInParentSeal int, missedRounds int64) {
+	sb.coreMu.RLock()
+	defer sb.coreMu.RUnlock()
+
+	countInParentSeal = -1
+
+	// Check the parent is not the genesis block.
+	number := child.Number().Uint64()
+	if number <= 1 {
+		return
+	}
+
+	childHeader := child.Header()
+	parentHeader := sb.chain.GetHeader(childHeader.ParentHash, number-1)
+
+	// Check validator in grandparent valset.
+	gpValSet := sb.getValidators(number-2, parentHeader.ParentHash)
+	gpValSetIndex, _ := gpValSet.GetByAddress(sb.ValidatorAddress())
+	if gpValSetIndex < 0 {
+		return
+	}
+	elected = true
+
+	// Now check if in the "parent seal" (used for downtime calcs, on the child block)
+	childExtra, err := types.ExtractIstanbulExtra(child.Header())
+	if err != nil {
+		return
+	}
+
+	pValSet := sb.getValidators(number-1, childHeader.ParentHash)
+	pValSetIndex, _ := pValSet.GetByAddress(sb.ValidatorAddress())
+	if pValSetIndex >= 0 {
+		inParentSeal = childExtra.ParentAggregatedSeal.Bitmap.Bit(pValSetIndex) != 0
+	}
+
+	missedRounds = childExtra.ParentAggregatedSeal.Round.Int64()
+	countInParentSeal = 0
+	for i := 0; i < gpValSet.Size(); i++ {
+		countInParentSeal += int(childExtra.ParentAggregatedSeal.Bitmap.Bit(i))
+	}
```

Comment:
Suggest calculating these earlier in this function so they are valid even if this node did not sign the block.